url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/ImagePersistence.java

```diff
@@ -0,0 +1,693 @@
+
+// Basic JDK classes
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.StringTokenizer;
+
+// Extended JDK classes
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+
+// VitralSDK classes
+import vsdk.toolkit.common.VSDK;
+import vsdk.toolkit.media.Image;
+import vsdk.toolkit.media.RGBImage;
+import vsdk.toolkit.media.RGBPixel;
+import vsdk.toolkit.media.RGBAImage;
+import vsdk.toolkit.render.awt.AwtRGBImageRenderer;
+import vsdk.toolkit.render.awt.AwtRGBAImageRenderer;
+import vsdk.toolkit.io.PersistenceElement;
+
+/**
+This class is a front end front which images of various formats can be
+exported and/or imported to/from files.
+
+@todo Does this implements a "Builder" design pattern??? A Factory design 
+      pattern... possibly some of that combined with a Facade design 
+      pattern?
+ */
+public class ImagePersistence extends PersistenceElement
+{
+    public static boolean nativeWarningGiven = false;
+
+    /**
+    Given the filename of an input data file which contains an image, this
+    method tries to recognize the file format and load the contents of it
+    to the image.
+
+    @todo Do not assume the file format only from the filename extension,
+    but trying to detect file headers.
+
+    @param imagen - The file respesenting the image
+    @return An RGBAImage entity that contains the image loaded in memory.
+
+    Will change:
+      - Choose a better name for this method
+      - Do not recieve a File, but a Stream of bytes
+    */
+    public static RGBAImage importRGBA(File inImageFd) throws ImageNotRecognizedException, Exception
+    {
+        String type = extractExtensionFromFile(inImageFd);
+        RGBAImage retImage = new RGBAImage();
+
+        //- Try optimized reading, if native library is available ---------
+        if ( type.equals("png") ) {
+            nativeWarningGiven = true;
+        }
+
+        //-----------------------------------------------------------------
+        if( type.equals("tga") ) {
+            TargaImage t = new TargaImage(inImageFd);
+            t.exportRGBA(retImage);
+            return retImage;
+        }
+        else if( type.equals("jpg") || type.equals("jpeg") ||
+                 type.equals("gif") || type.equals("png") )  {
+            BufferedImage bi = null;
+
+            // OLD SLOW METHOD, DO NOT USE!
+            //java.awt.Toolkit awtTools = java.awt.Toolkit.getDefaultToolkit();
+            //java.awt.Image image;
+            //image = awtTools.getImage(inImageFd.getAbsolutePath());
+            //bi = toBufferedImage(image);
+
+            try {
+                BufferedInputStream bis;
+                FileInputStream fis;
+
+                fis = new FileInputStream(inImageFd);
+                bis = new BufferedInputStream(fis);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *bis*. The resource is referenced at lines 97, 98. The resource is closed at lines 99, 100. There are other execution paths that do not contain closure statements, for example, when ImageIO.read throws an exception. Ensure *bis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).